loader/loader.go

```diff
@@ -1,5 +1,71 @@
 package loader
 
-// Loader represents loading process from a bitcoin client.
-type Loader interface {
+import (
+	"github.com/IlliniBlockchain/etl-bitcoin/client"
+	rpcclient "github.com/IlliniBlockchain/etl-bitcoin/client/rpc"
+	"github.com/IlliniBlockchain/etl-bitcoin/csv"
+	"github.com/IlliniBlockchain/etl-bitcoin/database"
+	"github.com/IlliniBlockchain/etl-bitcoin/types"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+)
+
+// General idea for loader:
+// Represents a single entire process of getting data from a source to a destination.
+// It takes in clients (RPC or DB) and implements the loading process for transporting
+// data between them.
+
+// LoaderOptions represents a map of arbitrary options when constructing a loader.
+type LoaderOptions map[string]interface{}
+
+// ClientDatabaseLoader represents any process of loading data from a
+// Client to a Database.
+type ClientDatabaseLoader interface {
+	Run()
+}
+
+type ClientDBLoaderConstructor func(client client.Client, db database.Database, opts LoaderOptions) (ClientDatabaseLoader, error)
+
+// DatabaseLoader represents any process of loading data from one type of
+// Database to another.
+type DatabaseLoader interface {
+}
+
+type DBLoaderConstructor func(srcDb, dstDb database.Database, opts LoaderOptions) (DBLoaderConstructor, error)
+
+// RPCCSVLoader represents the loading process for retrieving data from
+// a full node and saving it to disk as CSVs. Will implement
+// ClientDatabaseLoader.
+type RPCCSVLoader struct {
+	client *rpcclient.RPCClient
+	db     *csv.CSVDatabase
+}
+
+// CSVNeo4jLoader represents the loading process for uploading CSV data
+// into Neo4j. Will implement DatabaseLoader.
+type CSVNeo4jLoader struct {
+}
```

Comment:
Not sure I am following the specific Loaders for RPC->CSV and CSV->Neo4j because the whole point of the `Client` and `Database` interfaces is so that the Loader does not have to worry about the underlying implementations. So, rather you can more directly implement the loaders Client->Database and Database->database instead of creating individual structs for every implementation because they should all share the common functions defined by the `Client` and `Database` interfaces anyway.

So, maybe look at simplifying this to where you use the functions you have defined below of `BlockRangesToHashes`, `HashesToBlocks`, etc. to setup a pipeline that takes in a block range and a `DBTx` and uses that to get blocks and txs to call the [functions in `DBTx`](https://github.com/IlliniBlockchain/etl-bitcoin/blob/main/database/database.go#L23) as changes to the underlying database. Then at the end of the pipeline, we will have a filled `DBTx` and can call [`Commit`](https://github.com/IlliniBlockchain/etl-bitcoin/blob/main/database/database.go#L27) to store the changes.

Everything else looks great so far though.

Reply:
Also, I wouldn't worry about defining Database->Database loader for now, because specifically for Neo4j, there is a [command that we will run to bulk import CSVs into Neo4j](https://neo4j.com/developer/guide-import-csv/#batch-importer). It will be the `Neo4jCSVDatabase` job to format those CSVs correctly.

But that is an interesting idea for the future, had not considered defining loaders from database to database.